File: springboot/distribusi.stok.barang/src/main/java/be/distribusi/stok/barang/service/insert/InsertService.java
package be.distribusi.stok.barang.service.insert;

import be.distribusi.stok.barang.dto.insert.ReqInsertDTO;
import be.distribusi.stok.barang.dto.insert.ResInsertDTO;
import be.distribusi.stok.barang.exception.AppException;
import be.distribusi.stok.barang.repository.BarangRepository;
import be.distribusi.stok.barang.repository.ErrorRepository;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

@Service
@Slf4j
@RequiredArgsConstructor
public class InsertService {
    private final ErrorRepository errorConfig;
    private final BarangRepository barangRepository;

    public ResponseEntity<?> insertBarangService(@Valid ReqInsertDTO requestBody){
        // Buat objek LocalDateTime
        LocalDateTime now = LocalDateTime.now();
        // Format LocalDateTime menjadi String
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
        String loggerId = now.format(formatter);

        String kodeBarang = requestBody.getKodeBarang().toUpperCase();
        String namaBarang = requestBody.getNamaBarang().toUpperCase();
        BigDecimal hargaBeli = new BigDecimal(requestBody.getHargaBeli());
        BigDecimal hargaJual = new BigDecimal(requestBody.getHargaJual());
        int stokMasuk = Integer.parseInt(requestBody.getStokMasuk());
        int sisaStok = Integer.parseInt(requestBody.getStokMasuk());
        int stokKeluar = 0;

        // Cek apakah kode_barang sudah ada
        int isExist = barangRepository.selectByKode(kodeBarang);
        if (isExist == 0) {
            // Lakukan insert barang
            barangRepository.insertStokBarang(kodeBarang, namaBarang, hargaBeli, hargaJual, sisaStok, stokMasuk);

            // Mapping request ke responseDTO
            ResInsertDTO.insertDataDTO insertDataDTO = new ResInsertDTO.insertDataDTO();
            // Set data ke responseDTO
            insertDataDTO.setLoggerId(loggerId);
            insertDataDTO.setKodeBarang(kodeBarang);
            insertDataDTO.setNamaBarang(namaBarang);
            insertDataDTO.setHargaBeli(hargaBeli.toString());
            insertDataDTO.setHargaJual(hargaJual.toString());
            insertDataDTO.setSisaStok(Integer.toString(sisaStok));
            insertDataDTO.setStokMasuk(Integer.toString(stokMasuk));
            insertDataDTO.setStokKeluar(Integer.toString(stokKeluar));

            ResInsertDTO responseDTO = new ResInsertDTO();
            responseDTO.setData(insertDataDTO);
            responseDTO.setStatusCode("200");
            responseDTO.setStatusMsg("SUCCESS");
            responseDTO.setErrorMessage("");
            log.info("{} - End-to-End processing complete.", loggerId);
            // Return successful response
            return ResponseEntity.status(HttpStatus.OK).body(responseDTO);
        } else {
            // Kode barang sudah ada
            log.error("{} - KODE BARANG telah terdaftar", loggerId);
            String statusCode = errorConfig.getErrorCode("code_exist");
            String errorMessage = errorConfig.getErrorMessage("code_exist");
            throw new AppException(statusCode, errorMessage, HttpStatus.BAD_REQUEST, loggerId);
        }
    }
}

File: python/distribusi-stok-barang/services/insertService.py
from datetime import datetime
from decimal import Decimal
from models.databaseModels import SelectByCode, InsertStokBarang
from helpers.config import Config as config
from exception.appException import AppException
from loguru import logger as log
import json

def InsertDistribusiBarang(req, db):
    result = {
        "data":{},
        "status_code": "",
        "status_msg": "",
        "error_msg": "",
    }
    loggerId = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    log.info(f"{loggerId} - Start to processing insert service. Request: \n{json.dumps(dict(req))}")

    kodeBarang = req.kode_barang.upper()
    namaBarang = req.nama_barang.upper()
    hargaBeli = Decimal(req.harga_beli)
    hargaJual = Decimal(req.harga_jual)
    stokMasuk = int(req.stok_masuk)
    sisaStok = stokMasuk
    stokKeluar = 0

    isExist = SelectByCode(db, kodeBarang)

    if (isExist == True):
        log.info(f"{loggerId} - KODE BARANG telah terdaftar")
        raise AppException(loggerId, config.param_error['code_exist'][0], config.param_failed, config.param_error['code_exist'][1])
    
    InsertStokBarang(
        db,
        kodeBarang,
        namaBarang,
        hargaBeli,
        hargaJual,
        stokMasuk,
        sisaStok,
        stokKeluar
    ) 
    
    data = {
        "logger_id" : loggerId,
        "kode_barang": kodeBarang,
        "nama_barang": namaBarang,
        "harga_beli": str(hargaBeli),
        "harga_jual": str(hargaJual),
        "sisa_stok": str(sisaStok),
        "stok_keluar": str(stokKeluar)
    }
    result["data"] = data
    result["status_code"] = "200"
    result["status_msg"] = config.param_success
    result["error_msg"] = ""
    log.info(f"{loggerId} - End-to-End processing complete. Result: \n{json.dumps(dict(result))}")
    return result
        

    

